### PR TITLE
Taking Brute or Burn Damage Reduces Stuns Take Two

### PR DESCRIPTION
**Pull request by SkeletalElite**
## About The Pull Request
This is an attempt at an improvement over the changes in #75063

When you are in stam crit, the amount of brute/burn damage you take is tracked, once you have taken 25 damage, your stamina instantly regenerates and stam crit ends regardless of how long you have been in stam crit or have remaining on your stam crit. Only attacks that deal damage exceeding 5 add to this amount.

Immobilizes, Paralyzes, Stuns, and Knockdowns will be reduced by a number of seconds equal to .35x the damage taken. The previous iteration of this PR was .5x the damage taken, this has been reduced due to concerns that stuns were too ineffective. Particularly for xenos who would undo their stuns in a single slash, this is no longer the case and xenos now have a roughly 3 second window after slashing someone to extend their stun before it ends. Much like stamina, the damage must be greater than 5 to reduce stuns.

 Many which both stun and deal damage have also had their order of operations changed to ensure that they are not self reducing. 

In order for a stun effect that also deals damage to not self reduce, the stun effect should be placed after the damage effect. Note that this only applies to damage which uses the apply_damage proc and if something is directly calling adjustBruteLoss() or similar (like all chems AFAIK) it is not necessary as the reduction does not apply.
There may be a few more effects using apply_damage like this that need their order of operations changed to make sure they aren't self reducing, though, which I will fix as I become aware of them.

## Why It's Good For The Game
These improvements seek to address 3 major concerns which were brought to my attention in the previous PR.

The first is people helping each other evade arrest by beating each other. The changes to stam crit should address this issue, as you must deal at least 25 cumulative damage to someone to unstun them from stam crit. I believe this is enough damage that someone you are trying to murder will have an opportunity to escape or fight back. At the same time it's also high enough that it won't be a more effective method to help someone evade arrest than methods that already exist because 25 is not an insignificant amount of damage. 

The second concern was that effects that both deal damage and cause a stun at the same time could reduce themselves. In my testing, this only occurs on some stuns. While I'm certain I haven't fixed every stun that has this issue, it's a pretty easy fix and I can do it as I find more stuns that have this problem, so if you find a stun that's reducing itself, please let me know.

The third concern was addressed by the rework of how stamina damage second wind is handled. Previously if it was too agressive it was too easy to unstun people, but if it wasn't aggressive enough, you could still beat people to death while continuously extending stam crit by applying more stamina damage.  Because we now track the damage someone has taken since they entered stam crit, and end it instantly once a threshold is met this is no longer an issue.  This means stam crit cannot be extended in between damaging hits to keep someone stunned while murdering them. 

This should be acceptable with the head coder pronouncements regarding the balance of stun batons. Stun batons are still effective enough tools that average skill level security players can perform their jobs. They are also effective enough that stealth antagonists can still perform when ambusing people since someone who is ambushed is also easy enough to cuff. When it comes to stun batons this primarily nerfs them as a murderboning tool, since you cannot just baton someone twice, kill them. Then repeat on the next person you see. 

The remainder of this section is a copy paste from the previous PR:

Stuns have to be balanced around the fact that they are lethal, but at the same time there has to be effective methods in place to non lethally detain players. This creates a precarious situation to balance around because it's very difficult to make a stun good enough for security and non lethal detainment, but not so good it's one of the best ways to kill people.

At the very least, to kill someone off the back of a single stun with this change, you MUST cuff them first. The more steps there are between stunned and dead, the less we have to be concerned that a stun is too good as a way to kill people but also remain an effective tool for non lethal detainment.
## Changelog
:cl:
balance: Any stuns except those that knock you completely unconscious are reduced when you are attacked and take brute or burn damage. They are reduced by .35x the damage taken in seconds. 
balance: While in stamina crit, taking a cumulative 25 brute or burn damage from attacks will give you a second wind and cause you to regenerate your stamina. 
balance: Only attacks that deal greater than 5 damage will reduce stuns or count towards the cumulative damage to end stamina crit. 
/:cl:

### CODE REVIEW

**Comment by moocowswag** [comment]
I cant wait to see how confusing revs are gonna be for sec, they are fighting someone and stun baton them only for someone else to run up to the bad guy and toolbox him twice, then they both attack you.

**Comment by SkeletalElite** [comment]
> Imagine an assistant (rev) attacks you, you stun baton them and then a bystander "helps" you by attacking them with a toolbox twice then they both turn on you.

Toolboxes actually have 12 force and therefore you would need at least 3 hits, or 36 damage to your team mate to unstun them. There are much easier ways to sabotage an arrest than dealing 36 damage to your team mate like body throwing, shoving, or just dragging the dude being arrested off. If you had a weapon that dealt 13 damage and the person you are targeting had no armor you could unstun them in exactly 2 hits and 26 damage though, which is still not insignificant.

As for it happening accidentally, I think this is something people will learn to handle with experience.

**Comment by Vect0r2** [comment]
Seems really confusing for new player. Us OmniPlayers(tm) that keep us with this shit will know, but new players are gonna be confused when they hit somebody down, and they just get back up.

**Comment by carlarctg** [comment]
^ is true but i think it's still better than the alternative

**Comment by carlarctg** [comment]
i feel like this could be linked to determination

**Comment by YuliceTheDruid** [comment]
The beginning of Lethal Sec (Based)

**Comment by SkeletalElite** [comment]
> i feel like this could be linked to determination

This was one of my initial ideas but the problem with that that is that 1) low damage weapons rarely wound and 2) only extremely high damage weapons wound through armor.

This creates situations where armor actually makes you easier to kill because of how effective it is at preventing wounds

**Comment by SomeoneYouProbablyKnow** [comment]
> The beginning of Lethal Sec (Based)

The cause of lethal sec is not stun options or lack thereof (no matter what sec players tell you) it's the fact that you can bring people from the dead with wirecutters a blanket and a baton if you got the time. Being is not impactful so murder is just a long form stun. (Not that it's a bad thing i guess, i like actually playing the game instead getting killed forever)

**Comment by Mickyan** [comment]
This needs a feedback message when it triggers, either for the user or visible

**Comment by CRITAWAKETS** [comment]
this feels way too complicated for problems that could be solved in way simpler ways.

**Comment by moocowswag** [comment]
It should be tied to wounds (specifically when you get adrenaline/determination)

**Comment by SkeletalElite** [comment]
> It should be tied to wounds (specifically when you get adrenaline/determination)

The problem with that is that armor is so good at stopping wounds that people with armor would be easier to kill than people without it. You could easily harm baton a nukie a death for example but a literally naked guy has a decently high chance to get away.

**Comment by ATH1909** [comment]
> As for it happening accidentally, I think this is something people will learn to handle with experience. Edit: and it can create some funny slapstick momments too, having someone who's clueless but good intentioned accidentally help the bad guy

it'd be OOC slapstick, though, since there's not really a logical IC reason that hitting someone who's stunned would make them pop up with all of their stamina damage healed.

like, this isn't intuitive for a new player to know without being burnt AND being told by someone else that this very game-y mechanic exists and needs to be played around.

**Comment by SkeletalElite** [comment]
>  But some sort of visual indicator could go a long way, if you're interested i got a few ideas i can try and sprite

This could be cool. 

For now, I'll put a balloon alert, though.

**Comment by SomeoneYouProbablyKnow** [comment]
> it'd be OOC slapstick, though, since there's not really a logical IC reason that hitting someone who's stunned would make them pop up with all of their stamina damage healed.

What's the logical IC reasons for Stuns being as they are, last time i checked police batons don't cause paralysis in two hits? 
New players don't care about the exact math of stamina damage they just gotta know "if hit in stun, then less stun", they are gonna see how it works by simply existing in a round, just like they learn that cuffs prevent you from doing anything (which, is not how cuffs work logically)

The one thing i worry is lack of visual feedback but everyone's in agreement on that point.

**Comment by JupiterJaeden** [comment]
So, most people complain about the effectiveness of stun weaponry (particularly stun batons) in 1 on 1 fights, but this PR does absolutely nothing to nerf that. All it does is change is so that you have to cuff them before you kill them if you that is what you want to do - but it's still basically a "one click win". 

What it really does is massively nerf the effectiveness of stun weapons in GROUP fights - not only because of the ways this will certainly be cheesed, but also just because it's very difficult to stand still long enough to cuff someone while their buddies are still actively fighting you, then trying to damage them while dragging them cuffed mid-fight is going to slow you down massively. It's a BIG nerf to stun batons being used in lethal group fights - but group fights are one of the environments where regular non-nerfed stun batons are really not that OP, as mothblocks has previously pointed out. 

So basically, does nothing to nerf the stun baton in 1 on 1s, does a TON to nerf them in a group fight, which is really the opposite of the nerf that anyone was asking for. 

Sidenote: this is also a pretty serious buff to the freedom implant and changeling dissolve abilities, since previously the main counter to them was just beating people while they are stunned. I don't really think these two abilities needed a buff. 

If we are really going to nerf the baton, we should think about _why_ some people dislike it so much, and it's generally because they feel like the first baton hit is basically a guaranteed win. Why? Because they don't have enough time to react and generate any counterplay after the first hit. No brainer solution: increase the stun hit cooldown on the baton. Make it go from 2.5s to 4s like the telebaton. An extra 1.5s may not sound like a lot but I think it's enough to make a noticeable difference over many fights. And, if you don't have any items or abilities to help you generate some counterplay in that time, well, you _should_ just lose to the stun baton. Security should have a reliable melee weapon against unarmed shitters and the like. 

Also, maybe change the baton's status effect to something like knockdown in every way but without the ability to be shovestun paralyzed, because otherwise the added time doesn't really mean much if you can just click one more time to paralyze anyways.

**Comment by Comxy** [comment]
> While in stamina crit, taking a cumulative 25 brute or burn damage from attacks will give you a second wind and cause you to regenerate your stamina.

> Only attacks that deal greater than 5 damage will reduce stuns or count towards the cumulative damage to end stamina crit.

I really do not like these special thresholds that and that the reduction is .35x specifically. Is it not better to turn it into a mathematical formula that seems balanced, the more damage the more reduction takes place.

**Comment by SkeletalElite** [comment]
> I really do not like these special thresholds that and that the reduction is .35x specifically. Is it not better to turn it into a mathematical formula that seems balanced, the more damage the more reduction takes place.

If you have suggestions for a formula, suggest away

**Comment by moocowswag** [comment]
> So, most people complain about the effectiveness of stun weaponry (particularly stun batons) in 1 on 1 fights, but this PR does absolutely nothing to nerf that. 

Ya this pr has alot of entertaining problems coming with it and no solutions.

At the end of the day, in 1v1 stam combat is always gonna be superior as long as we value retaining the ability to arrest someone nonlethally despite them being a lethal threat that wants to fight to the death.

**Comment by Pickle-Coding** [comment]
This just simply doesn't make any sense. The stamina system already suffers from being very weird and unnatural. This makes it even worse.

This adds a lot of extra complexity that is very unnatural, and all it does is make people cuff their target before executing them. This complexity cost definitly outweighs the rest of the gameplay impact this is going to have.

**Comment by OverwatchVoice** [comment]
I'm all for nerfing stun batons in some way, but this just isn't the way to go.
Besides, all this will change is ensuring people will first get cuffed before getting the shit beaten out of them.

**Comment by Overuseddart** [comment]
Can’t wait for: I only shot him because I wanted him to get up faster ban appeal
All jokes aside this might be cool to see. (Shooting your nukie teammates would be a viable strategy)

**Comment by SkeletalElite** [comment]
> are we banking on people screaming help with stun nerfs or what?

The point is to slowdown stuns as a murder tool, such that if you want a fast kill, you should use a killing tool and not a stunning tool. Fights are more interesting when there is a back and forth engagement rather than first person to click other with stun wins on the spot. Of course, if you follow up a stun with cuffs, it is still possible to stun someone and then kill them. However, it takes longer, needs more preparation, and is much less feasible as a way to win when outnumbered.

It encourages players to cooperate rather than be lone wolves (people interacting in my roleplaying game?!), antags to be stealthy by making it harder to win fights outnumbered, and to use a variety of tools for different situations instead of the same old baton solving 90% of problems. As for security, when it comes to dealing with groups, they should be using teamwork themselves, as well as flashbangs, tear gas, riot armor, and (if necessary) lethals to deal with groups. The pepper spray backpack could also use a buff to not just be a more spacious shittier version of the pepper spray in your belt too, but that's outside the scope of my PR.

The baton still does what it needs to do, act as an effective way for security to arrest individuals non lethally and stealth antags to ambush lone players, but it's not a solve nearly all trump card for melee combat. If you're trying to kill someone, especially, in a group fight or if they are a dangerous individual who can't be contained by cuffs, something that directly kills is probably preferable.  That doesn't make the baton useless, though, if you're trying to non lethally arrrest someone (even if its so you can kill them later) the stun baton is still the best and most consistent tool for that. 

If the abundance of coils is an issue, we could make cable cuffs single-use like zip ties and increase the number of coils needed to make them, increase their size so they can't be put in a cardboard box, and add a short crafting time. That way more of them are required, or you have to get your hands on a pair of handcuffs, which are rarer (and more suspicious).

tl;dr:  This PR just takes a step towards a game where stun combat is not the most effective form of fighting, and that will make other (lethal) weapons more desirable by comparison, but in one or way or another it's making stuns worse for _everyone_ and even if stuns are used just little less as the primary form of winning fights, that's a win in my book.

**Comment by ATH1909** [comment]
If the goal here is to reduce the effectiveness of stun batons in group fights... what about just decreasing the default duration of staminacrit? Like, down from 10 seconds to, say, 5 or 6. 2-3 seconds of leeway should be enough time to whip out cuffs in isolated conditions, but not when you're being jostled around by an active fight.

**Comment by OverwatchVoice** [comment]
If I'm not mistaken, cable cuffs are already single use, but that's not the issue.

The time required to cuff someone if you'd want kill them is a miniscule difference, one which most people stunning to kill do anyways.

**Comment by ExcessiveUseOfCobblestone** [comment]
Sorry I'm arriving late to the party a bit, this is pretty much for group stunning?

I'd think you'd want it in reverse, groups that clobber on you gives little room for an escape but ideally you'd want some sort of mitigation from some guy getting a single baton hit on you to get pretty much free reign unless he continued to follow up with nonlethals (aka not hitting you with lethals)

Is there some sort of design approach that's been written for the topic you're trying to achieve or is this just a good spaghetti test pr?

**Comment by SkeletalElite** [comment]
>is there some sort of design approach that's been written for the topic you're trying to achieve or is this just a good spaghetti test pr?

No, the PR only attemptd to be compliant with the headcoder pronouncements regarding stun batons and stun combat, but there is no design otherwise and is just an interesting idea I had to try to shake up fighting in a way that could be more fun.
